docs/cvl/builtin.md

```diff
@@ -1,113 +1,179 @@
 Built-in Rules
 ==============
-The Certora Prover has built-in general-purpose rules targeted at finding known vulnerabilities.
-These rules can be verified on a contract out-of-the-box.
 
-Users can add built-in rules and other rule types in the same spec file.
+The Certora Prover has built-in general-purpose rules targeted at finding common
+vulnerabilities.  These rules can be verified on a contract without writing any
+contract-specific rules.
 
+Built-in rules can be included in any spec file by writing `use builtin rule
+<rule-name>;`.  This document describes the available built-in rules.
+
+```{contents}
+```
 
 Syntax
 ------
-The syntax for using the built-in rule identifier
+
+The syntax for rules is given by the following [EBNF grammar](syntax):
 
 ```
-built-in-rule ::= "use" "builtin" "rule" id ";"
-```
+built_in_rule ::= "use" "builtin" "rule" built_in_rule_name ";"
 
-Example
--------
-In order to run the built-in rule `msgValueInLoopRule` add to the spec file the line
+built_in_rule_name ::=
+    | "msgValueInLoopRule"
+    | "hasDelegateCalls"
+    | "sanity"
+    | "deepSanity"
+```
 
-`use builtin rule msgValueInLoopRule;`
+(built-in-msg-value-in-loop)=
+Bad loop detection &mdash; `msgValueInLoopRule`
+-----------------------------------------------
 
-The Names of Currently Implemented Rules:
------------------------------------------
-- `msgValueInLoopRule` – checks for occurrences of `msg.value` and delegate calls in loops.
-- `hasDelegateCalls` - checks for delegate calls anywhere in the contract.
-- `sanity`
-- `deepSanity`
+Loops that [use `msg.value`][msg-value-vulnerability] or [make delegate
+calls][delegatecall-vulnerability] are a well-known source of security
+vulnerabilities.
 
+[msg-value-vulnerability]: https://trustchain.medium.com/ethereum-msg-value-reuse-vulnerability-5afd0aa2bcef
+[delegatecall-vulnerability]: https://blog.trailofbits.com/2021/12/16/detecting-miso-and-opyns-msg-value-reuse-vulnerability-with-slither/
 
-## Sanity
+```{todo}
+are these the best citations?
+```
 
-Enable with:
+The `msgValueInLoopRule` detects these anti-patterns.  It can be enabled by
+including
 ```cvl
-use builtin rule sanity;
+use builtin rule msgValueInLoopRule;
 ```
+in a spec file.  The rule will fail on any functions that can make delegate
+calls or access `msg.value` inside a loop.
 
-The sanity rule acts as one of our main methods to set up a new code base for verification. It serves two needs:
+```{todo}
+I imagine it's actually on any functions that recursively call any functions that do this?
+And probably also loops that call functions that recursively access `msg.value`?
+Not sure if this needs more clarification or not.
+```
```

Comment:
yes it is but I am not sure if it's obvious or not. Maybe we should only add a clarification or in the sentence above

Reply:
It is. We have tests for this and I checked it again.